File: workspace/solutions/ecmascript/src/s1/s8/s5/s6/s_1856_00.js
/*
 * @lc app=leetcode.cn id=1856 lang=javascript
 *
 * [1856] 子数组最小乘积的最大值
 */

// @lc code=start
const MOD = BigInt(1e9 + 7);
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxSumMinProduct = function (nums) {
    const n = nums.length;
    const left = Array.from({ length: n });
    const right = Array.from({ length: n });
    const stack = [];

    for (let i = 0; i < n; i++) {
        // @ts-ignore
        while (stack.length && nums[stack.at(-1)] >= nums[i]) {
            stack.pop();
        }
        left[i] = stack.at(-1) ?? -1;
        stack.push(i);
    }

    stack.length = 0;
    for (let i = n - 1; i >= 0; i--) {
        // @ts-ignore
        while (stack.length && nums[stack.at(-1)] > nums[i]) {
            stack.pop();
        }
        right[i] = stack.at(-1) ?? n;
        stack.push(i);
    }

    const prefix_sums = Array.from({ length: n + 1 });
    prefix_sums[0] = 0;
    for (let i = 0; i < n; i++) {
        prefix_sums[i + 1] = prefix_sums[i] + nums[i];
    }

    const result = nums.reduce((result, num, i) => {
        const product = BigInt(prefix_sums[right[i]] - prefix_sums[left[i] + 1]) * BigInt(num);
        return result > product ? result : product;
    }, 0n);
    return Number(result % MOD);
};
// @lc code=end

export default maxSumMinProduct;

File: workspace/solutions/rust/src/s1/s8/s5/s6/s_1856_00.rs
/*
 * @lc app=leetcode.cn id=1856 lang=rust
 *
 * [1856] 子数组最小乘积的最大值
 */

pub struct Solution {}

// @lc code=start
const MOD: i64 = 1_000_000_007;

impl Solution {
    pub fn max_sum_min_product(nums: Vec<i32>) -> i32 {
        /* 使用单调栈计算数组每个元素的作为最小值的最大子数组 */
        let n = nums.len();
        /* left_index[i] 表示数组元素 nums[i] 左侧第一个 < nums[i] 的坐标 */
        let mut left_index = vec![-1; n];
        /* right_index[i] 表示数组元素 nums[i] 右侧第一个 <= nums[i] 的坐标 */
        let mut right_index = vec![n as i32; n];
        /* 单调递增栈 */
        let mut stack = std::collections::VecDeque::new();
        /* nums 的前缀和 */
        let mut prefix_sums = vec![0_i64; n + 1];

        /* 计算子数组左边界 */
        for (i, &num) in nums.iter().enumerate() {
            while !stack.is_empty() && nums[*stack.back().unwrap()] >= num {
                stack.pop_back();
            }
            if let Some(&top) = stack.back() {
                left_index[i] = top as i32;
            }
            stack.push_back(i);
        }

        /* 计算子数组右边界 */
        stack.clear();
        for (i, &num) in nums.iter().rev().enumerate() {
            let j = n - i - 1;
            while !stack.is_empty() && nums[*stack.back().unwrap()] > num {
                stack.pop_back();
            }
            if let Some(&top) = stack.back() {
                right_index[j] = top as i32;
            }
            stack.push_back(j);
        }

        /* 计算数组前缀和 */
        for (i, &num) in nums.iter().enumerate() {
            prefix_sums[i + 1] = prefix_sums[i] + (num as i64);
        }

        /* 计算每个元素对应的子数组最小乘积的最大值 */
        let result = nums.iter().enumerate().fold(0_i64, |result, (i, &num)| -> i64 {
            let left = left_index[i];
            let right = right_index[i];
            let sum = prefix_sums[right as usize] - prefix_sums[(left + 1) as usize];
            return result.max(sum * (num as i64));
        });

        (result % MOD) as i32
    }
}
// @lc code=end

pub const SOLUTION: super::TSolution = Solution::max_sum_min_product;
